core/node/vm_runner/src/process.rs

```diff
@@ -0,0 +1,177 @@
+
+/// VM runner represents a logic layer of L1 batch / L2 block processing flow akin to that of state
+/// keeper. The difference is that VM runner is designed to be run on batches/blocks that have
+/// already been processed by state keeper but still require some extra handling as regulated by
+/// [`OutputHandlerFactory`].
+///
+/// It's responsible for taking unprocessed data from the [`VmRunnerIo`], feeding it into
+/// [`BatchExecutor`] and calling [`OutputHandlerFactory`] on the result of the execution (batch
+/// execution state in the [`UpdatesManager`]).
+///
+/// You can think of VM runner as a concurrent processor of a continuous stream of newly committed
+/// batches/blocks.
+#[derive(Debug)]
+pub struct VmRunner {
+    pool: ConnectionPool<Core>,
+    io: Box<dyn VmRunnerIo>,
+    storage: Arc<dyn ReadStorageFactory>,
+    loader: Arc<dyn StorageLoader>,
+    output_handler_factory: Box<dyn OutputHandlerFactory>,
+    batch_processor: Box<dyn BatchExecutor>,
+}
+
+impl VmRunner {
+    /// Initializes VM runner with its constituents. In order to make VM runner concurrent each
+    /// parameter here needs to support concurrent execution mode. See
+    /// [`ConcurrentOutputHandlerFactory`], [`VmRunnerStorage`].
+    ///
+    /// Caller is expected to provide a component-specific implementation of [`VmRunnerIo`] and
+    /// an underlying implementation of [`OutputHandlerFactory`].
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        io: Box<dyn VmRunnerIo>,
+        // Normally `storage` and `loader` would be the same object, but it is awkward to deal with
+        // them as such due to Rust's limitations on upcasting coercion. See
+        // https://github.com/rust-lang/rust/issues/65991.
+        storage: Arc<dyn ReadStorageFactory>,
+        loader: Arc<dyn StorageLoader>,
+        output_handler_factory: Box<dyn OutputHandlerFactory>,
+        batch_processor: Box<dyn BatchExecutor>,
+    ) -> Self {
+        Self {
+            pool,
+            io,
+            storage,
+            loader,
+            output_handler_factory,
+            batch_processor,
+        }
+    }
+
+    /// Consumes VM runner to execute a loop that continuously pulls data from [`VmRunnerIo`] and
+    /// processes it.
+    pub async fn run(mut self, stop_receiver: &watch::Receiver<bool>) -> anyhow::Result<()> {
+        const SLEEP_INTERVAL: Duration = Duration::from_millis(50);
+
+        let mut handles = Vec::<JoinHandle<anyhow::Result<()>>>::new();
+        let mut next_batch = self
+            .io
+            .latest_processed_batch(&mut self.pool.connection().await?)
+            .await?
+            + 1;
+        loop {
+            let mut retained_handles = Vec::new();
+            for handle in handles {
+                if handle.is_finished() {
+                    handle
+                        .await
+                        .context("Failed to join a task to process a batch")?
```

Comment:
- Nit: The outer error is most probably a panic (it could also be a task getting aborted, but unless you call `JoinHandle::abort()`, this shouldn't happen). Thus, it could make sense to rephrase to "Processing a batch panicked".
- Nit: Is the batch number obvious from the source error? Otherwise, it'd be relatively easy to add it by using `Vec<(L1BatchNumber, JoinHandle<_>)>` or `HashMap<L1BatchNumber, _>`.